presto-main/src/main/java/com/facebook/presto/operator/aggregation/AccumulatorCompiler.java

```diff
@@ -553,33 +596,69 @@ private static void generateAddIntermediateAsCombine(
+                block.add(columnBlock);
+            }
+        }
+
         Variable position = scope.declareVariable(int.class, "position");
+        for (int i = 0; i < stateCount; i++) {
+            FieldDefinition stateFactoryField = stateFieldAndMetadatas.get(i).getStateFactoryField();
+            body.comment(format("scratchState_%s = stateFactory[%s].createSingleState();", i, i))
+                    .append(thisVariable.getField(stateFactoryField))
+                    .invokeInterface(AccumulatorStateFactory.class, "createSingleState", Object.class)
+                    .checkCast(scratchState.get(i).getType())
+                    .putVariable(scratchState.get(i));
+        }
 
-        body.comment("scratchState = stateFactory.createSingleState();")
-                .append(thisVariable.getField(stateFactoryField))
-                .invokeInterface(AccumulatorStateFactory.class, "createSingleState", Object.class)
-                .checkCast(scratchState.getType())
-                .putVariable(scratchState);
+        List<FieldDefinition> stateFields = stateFieldAndMetadatas.stream()
+                .map(StateFieldAndMetadata::getStateField)
+                .collect(toImmutableList());
 
         if (grouped) {
-            generateEnsureCapacity(scope, stateField, body);
+            generateEnsureCapacity(scope, stateFields, body);
         }
 
         BytecodeBlock loopBody = new BytecodeBlock();
 
         if (grouped) {
             Variable groupIdsBlock = scope.getVariable("groupIdsBlock");
-            loopBody.append(thisVariable.getField(stateField).invoke("setGroupId", void.class, groupIdsBlock.invoke("getGroupId", long.class, position)));
+            for (FieldDefinition stateField : stateFields) {
+                loopBody.append(thisVariable.getField(stateField).invoke("setGroupId", void.class, groupIdsBlock.invoke("getGroupId", long.class, position)));
+            }
         }
 
-        loopBody.append(thisVariable.getField(stateSerializerField).invoke("deserialize", void.class, block, position, scratchState.cast(Object.class)));
+        for (int i = 0; i < stateCount; i++) {
+            FieldDefinition stateSerializerField = stateFieldAndMetadatas.get(i).getStateSerializerField();
+            loopBody.append(thisVariable.getField(stateSerializerField).invoke("deserialize", void.class, block.get(i), position, scratchState.get(i).cast(Object.class)));
+        }
 
-        loopBody.comment("combine(state, scratchState)")
-                .append(thisVariable.getField(stateField))
-                .append(scratchState)
-                .append(invoke(callSiteBinder.bind(combineFunction), "combine"));
+        loopBody.comment("combine(state_0, state_1, ... scratchState_0, scratchState_1, ...)");
+        for (FieldDefinition stateField : stateFields) {
+            loopBody.append(thisVariable.getField(stateField));
+        }
```

Comment:
I think the readability would improve if the three `for`-loops are combined into one. I'm not sure though.

Reply:
```
        loopBody.comment("combine(state_0, state_1, ... scratchState_0, scratchState_1, ...)");
        for (FieldDefinition stateField : stateFields) {
            if (grouped) {
                loopBody.append(thisVariable.getField(stateField).invoke("setGroupId", void.class, groupIdsBlock.invoke("getGroupId", long.class, position)));
            }
            loopBody.append(thisVariable.getField(stateField));
        }
        for (int i = 0; i < stateCount; i++) {
            FieldDefinition stateSerializerField = stateFieldAndDescriptors.get(i).getStateSerializerField();
            loopBody.append(thisVariable.getField(stateSerializerField).invoke("deserialize", void.class, block.get(i), position, scratchStates.get(i).cast(Object.class)));
            loopBody.append(scratchStates.get(i));
        }